mmaction/models/heads/ssn_head.py

```diff
@@ -0,0 +1,439 @@
+            self.activity_slice = slice(0, self.activity_score_len)
+
+        self.complete_slice = slice(
+            self.activity_slice.stop, self.activity_slice.stop +
+            self.complete_score_len * self.feat_multiplier)
+        self.reg_slice = slice(
+            self.complete_slice.stop, self.complete_slice.stop +
+            self.reg_score_len * self.feat_multiplier)
+
+    def forward(self, x, proposal_ticks, scale_factors):
+        """Defines the computation performed at every call.
+
+        Args:
+            x (torch.Tensor): The input data.
+            proposal_ticks (list): Ticks of proposals to be STPP.
+            scale_factors (list): Ratios of the effective sampling lengths
+                to augmented lengths.
+
+        Returns:
+            tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+                out_activity_scores (torch.Tensor): Activity scores
+                out_complete_scores (torch.Tensor): Completeness scores.
+                out_reg_scores (torch.Tensor): Regression scores.
+        """
+        assert x.size(1) == self.feat_dim
+        num_ticks = proposal_ticks.size(0)
+
+        out_activity_scores = torch.zeros((num_ticks, self.activity_score_len),
+                                          dtype=x.dtype)
+        raw_activity_scores = x[:, self.activity_slice]
+
+        out_complete_scores = torch.zeros((num_ticks, self.complete_score_len),
+                                          dtype=x.dtype)
+        raw_complete_scores = x[:, self.complete_slice]
+
+        if self.with_regression:
+            out_reg_scores = torch.zeros((num_ticks, self.reg_score_len),
+                                         dtype=x.dtype)
+            raw_reg_scores = x[:, self.reg_slice]
+        else:
+            out_reg_scores = None
+            raw_reg_scores = None
+
+        def pyramids_pooling(out_scores, index, raw_scores, ticks,
+                             scale_factors, score_len, stpp_stage):
+            """Perform pyramids pooling.
+
+            Args:
+                out_scores (torch.Tensor): Scores to be returned.
+                index (int): Index of output scores.
+                raw_scores (torch.Tensor): Raw scores before STPP.
+                ticks (int): Ticks of raw scores.
+                scale_factors (list): Ratios of the effective sampling lengths
+                    to augmented lengths.
+                score_len (int): Length of the score.
+                stpp_stage (tuple): Config of STPP.
+            """
+            offset = 0
+            for stage_idx, stage_cfg in enumerate(stpp_stage):
+                if stage_idx == 0:
+                    scale_factor = scale_factors[0]
+                elif stage_idx == len(stpp_stage) - 1:
+                    scale_factor = scale_factors[1]
+                else:
+                    scale_factor = 1.0
+
+                sum_parts = sum(stage_cfg)
+                tick_left = float(ticks[stage_idx])
```

Comment:
is it safe to remove `float`?

Reply:
Maybe it is better to add float to `tick_right`. If `tick_right` is `torch.Tensor(3, dtype=np.int32)`, then `torch.arange(tick_left, tick_right+1e-5, num_parts)` will not work.

Reply:
`tick_right+1e-5` -> `torch.Tensor(3, dtype=np.int32)`